tests/JustSaying.IntegrationTests/Fluent/Configuration/WhenUsingACustomSubjectProvider.cs

```diff
@@ -0,0 +1,87 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using JustSaying.IntegrationTests.Fluent;
+using JustSaying.IntegrationTests.Fluent.Subscribing;
+using JustSaying.Messaging.MessageHandling;
+using JustSaying.Messaging.MessageSerialization;
+using JustSaying.TestingFramework;
+using Microsoft.Extensions.DependencyInjection;
+using Newtonsoft.Json;
+using Shouldly;
+using Xunit.Abstractions;
+
+namespace JustSaying.Fluent.Configuration
+{
+    public class WhenUsingACustomSubjectProvider : IntegrationTestBase
+    {
+        public WhenUsingACustomSubjectProvider(ITestOutputHelper outputHelper) : base(outputHelper)
+        { }
+
+        [AwsFact]
+        public async Task ThenItIsUsed()
+        {
+            // Arrange
+            var handler = new InspectableHandler<SimpleMessage>();
+            var accessor = new RecordingMessageContextAccessor(new MessageContextAccessor());
+
+            var subject = Guid.NewGuid().ToString();
+            var subjectProvider = new ConstantSubjectProvider(subject);
+
+            var services = GivenJustSaying()
+                .ConfigureJustSaying((builder) =>
+                    builder.WithLoopbackTopic<SimpleMessage>(UniqueName))
+                .ConfigureJustSaying(builder =>
+                    builder.Services(s => s.WithMessageContextAccessor(() => accessor)))
+                .ConfigureJustSaying((builder) =>
+                    builder.Messaging(m =>
+                        m.WithMessageSubjectProvider(subjectProvider)))
+                .AddSingleton<IHandlerAsync<SimpleMessage>>(handler);
+
+            var id = Guid.NewGuid();
+            var message = new SimpleMessage()
+            {
+                Id = id
+            };
+
+            var cts = new CancellationTokenSource(TimeSpan.FromSeconds(1));
+
+            await WhenAsync(
+                services,
+                async (publisher, listener, cancellationToken) =>
+                {
+                    await listener.StartAsync(cancellationToken);
+                    await publisher.StartAsync(cancellationToken);
+
+                    // Let's send an OrderPlaced, but the subject will be SimpleMessage
+                    // because of the custom subject provider
+                    await publisher.PublishAsync(message, cancellationToken);
+
+                    await cts.Token.WaitForCancellation();
+
+                    handler.ReceivedMessages.ShouldHaveSingleItem().Id.ShouldBe(id);
+                    var context = accessor.ValuesWritten.ShouldHaveSingleItem();
+                    dynamic json = JsonConvert.DeserializeObject(context.Message.Body);
+                    string subject = json.Subject;
+                    subject.ShouldBe(subject);
+                });
+
+            cts.Dispose();
+        }
+    }
+
+    public class ConstantSubjectProvider : IMessageSubjectProvider
```

Comment:
I'd make this nested if it's specific to this one test.

Reply:
Done!